# Confirmatory Factor Analysis

This notebook employs Confirmatory Factor Analysis (CFA) to validate our research in line with three central constructs. These constructs serve as the foundation for our hypotheses:

**Eco-labeling Perception**: We assess if multiple eco-labeling schemes lead to consumer confusion and hinder awareness of eco-friendly products (Hypothesis 1).

**Consumer Trust**: We investigate the trust consumers place in products bearing the KEBS label concerning quality and safety (Hypothesis 2).

**Consumer Attitudes**: We analyze consumer attitudes concerning the KEBS label and its association with government oversight and product choice (Hypothesis 3).

In [27]:
import warnings
import pandas as pd
import numpy as np
import semopy

data = pd.read_csv('cfa.csv')
# To filter out all warnings
warnings.filterwarnings("ignore")

In [28]:
data

Unnamed: 0 What is your age?  What is your Gender?  \
0             0             18-24                   1.0   
1             1             18-24                   1.0   
2             2             55-64                   0.0   
3             3             25-34                   0.0   
4             4             35-44                   0.0   
..          ...               ...                   ...   
110         110             18-24                   1.0   
111         111             18-24                   0.0   
112         112             18-24                   NaN   
113         113             35-44                   1.0   
114         114             25-34                   0.0   

     Which region in Kenya do you currently reside in?  \
0                                                    5   
1                                                    4   
2                                                    3   
3                                                    5   
4                                                    2   
..                                                 ...   
110                                                  5   
111                                                  5   
112                                                  5   
113                                                  3   
114                                                  3   

     How aware are you about the environmental issues?(1 being the lowest,5 being the highest)  \
0                                                    5                                           
1                                                    5                                           
2                                                    5                                           
3                                                    4                                           
4                                                    4                                           
..                                                 ...                                           
110                                                  4                                           
111                                                  4                                           
112                                                  5                                           
113                                                  3                                           
114                                                  4                                           

     Do you actively engage in environment-friendly practices in your daily life?  \
0                                                  0.0                              
1                                                  1.0                              
2                                                  1.0                              
3                                                  1.0                              
4                                                  1.0                              
..                                                 ...                              
110                                                1.0                              
111                                                1.0                              
112                                                1.0                              
113                                                1.0                              
114                                                1.0                              

     I am confident in my ability to distinguish between different eco-labels when shopping for products.  \
0                                                    5                                                      
1                                                    4                                                      
2                                                    1                    

In [29]:
###Select data for the Eco-label Perception construct
d1 = data[['I am confident in my ability to distinguish between different eco-labels when shopping for products.',
         'Have you ever encountered multiple eco-labels on similar products?',
         'How often do you find it challenging to understand the environmental implications of a product based on its eco-label?']]
d1

I am confident in my ability to distinguish between different eco-labels when shopping for products.  \
0                                                    5                                                      
1                                                    4                                                      
2                                                    1                                                      
3                                                    3                                                      
4                                                    3                                                      
..                                                 ...                                                      
110                                                  3                                                      
111                                                  3                                                      
112                                                  5                                                      
113                                                  2                                                      
114                                                  4                                                      

     Have you ever encountered multiple eco-labels on similar products?  \
0                                                    1                    
1                                                    0                    
2                                                    0                    
3                                                    1                    
4                                                    1                    
..                                                 ...                    
110                                                  0                    
111                                                  0                    
112                                                  1                    
113                                                  1                    
114                                                  1                    

     How often do you find it challenging to understand the environmental implications of a product based on its eco-label?  
0                                                    5                                                                       
1                                                    4                                                                       
2                                                    5                                                                       
3                                                    4                                                                       
4                                                    3                                                                       
..                                                 ...                                                                       
110                                                  4                                                                       
111                                                  3                                                                       
112                                                  2                                                                       
113                                                  3                                                                       
114                                                  4                                                                       

[115 rows x 3 columns]

In [30]:
### Select data for the Trust construct
d2 = data[['How much trust do you place in products that bear the KEBS label when it comes to quality and safety?','Do you believe that products with the KEBS label are more reliable and safer than those without it?']]
d2

How much trust do you place in products that bear the KEBS label when it comes to quality and safety?  \
0                                                  4.0                                                       
1                                                  NaN                                                       
2                                                  NaN                                                       
3                                                  3.0                                                       
4                                                  3.0                                                       
..                                                 ...                                                       
110                                                5.0                                                       
111                                                5.0                                                       
112                                                5.0                                                       
113                                                5.0                                                       
114                                                5.0                                                       

     Do you believe that products with the KEBS label are more reliable and safer than those without it?  
0                                                    1                                                    
1                                                    1                                                    
2                                                    1                                                    
3                                                    0                                                    
4                                                    0                                                    
..                                                 ...                                                    
110                                                  1                                                    
111                                                  1                                                    
112                                                  1                                                    
113                                                  1                                                    
114                                                  1                                                    

[115 rows x 2 columns]

In [31]:
#### Select data to test the Attitude Construct
d3 = data[['Have you intentionally purchased products with the KEBS label because of its perceived quality and safety standards?','To what extent do you associate the KEBS label with government oversight and regulation?','How likely are you to choose a product with the KEBS label over a similar product without it?']]
d3

Have you intentionally purchased products with the KEBS label because of its perceived quality and safety standards?  \
0                                                    1                                                                      
1                                                    1                                                                      
2                                                    1                                                                      
3                                                    0                                                                      
4                                                    0                                                                      
..                                                 ...                                                                      
110                                                  1                                                                      
111                                                  1                                                                      
112                                                  1                                                                      
113                                                  1                                                                      
114                                                  1                                                                      

     To what extent do you associate the KEBS label with government oversight and regulation?  \
0                                                    3                                          
1                                                    5                                          
2                                                    5                                          
3                                                    3                                          
4                                                    4                                          
..                                                 ...                                          
110                                                  5                                          
111                                                  5                                          
112                                                  4                                          
113                                                  4                                          
114                                                  5                                          

     How likely are you to choose a product with the KEBS label over a similar product without it?  
0                                                    4                                              
1                                                    5                                              
2                                                    5                                              
3                                                    3                                              
4                                                    4                                              
..                                                 ...                                              
110                                                  5                                              
111                                                  5                                              
112                                                  5                                              
113                                                  5                                              
114                                                  5                                              

[115 rows x 3 columns]

# Ecolabel Perception

In [32]:
#Rename the columns to use them easily
d1.rename({
    'I am confident in my ability to distinguish between different eco-labels when shopping for products.': 'Confidence',
    'Have you ever encountered multiple eco-labels on similar products?': 'Encounter_Multiple_Labels',
    'How often do you find it challenging to understand the environmental implications of a product based on its eco-label?': 'Challenging_Understanding'
}, axis=1, inplace=True)
d1

Confidence  Encounter_Multiple_Labels  Challenging_Understanding
0             5                          1                          5
1             4                          0                          4
2             1                          0                          5
3             3                          1                          4
4             3                          1                          3
..          ...                        ...                        ...
110           3                          0                          4
111           3                          0                          3
112           5                          1                          2
113           2                          1                          3
114           4                          1                          4

[115 rows x 3 columns]

In [33]:
#Carry out CFA
import semopy

model = '''
    # Measurement model
    f1 =~ Confidence
    f2 =~ Encounter_Multiple_Labels
    f3 =~ Challenging_Understanding
'''
model = semopy.Model(model)
res =model.fit(d1)
print(res)
 

Name of objective: MLW
Optimization method: SLSQP
Optimization successful.
Optimization terminated successfully
Objective value: 0.000
Number of iterations: 15
Params: 0.570 0.937 0.138 0.346 0.095 -0.103 0.111 -0.094 0.223


# Intepretation
In Confirmatory Factor Analysis (CFA), the factor loadings represent the relationships between latent factors (in this case, your Eco-label Perception Construct) and the observed variables (the three questions). The factor loadings are standardized coefficients that indicate how much variance in the observed variables is explained by the underlying latent construct. 

"I am confident in my ability to distinguish between different eco-labels when shopping for products."

Factor Loading: 0.570
Interpretation: The factor loading of 0.570 suggests that this question is moderately positively associated with the Eco-label Perception Construct. In other words, individuals who report higher confidence in distinguishing between eco-labels are more likely to score higher on the Eco-label Perception Construct. This question contributes significantly to the measurement of the Eco-label Perception Construct.
"Have you ever encountered multiple eco-labels on similar products?"

Factor Loading: 0.937
Interpretation: The factor loading of 0.937 indicates a strong positive association between encountering multiple eco-labels and the Eco-label Perception Construct. Individuals who have encountered multiple eco-labels on similar products are more likely to have a higher score on the Eco-label Perception Construct. This question is a strong indicator of the construct.
"How often do you find it challenging to understand the environmental implications of a product based on its eco-label?"

Factor Loading: 0.138
Interpretation: The factor loading of 0.138 suggests that this question is weakly positively associated with the Eco-label Perception Construct. It contributes to the measurement of the construct, but to a lesser extent compared to the other two questions. Individuals who find it challenging to understand eco-labels tend to have a slightly higher score on the Eco-label Perception Construct

## Trust Construct

In [34]:
# Define a dictionary to map the old column names to new short names
column_mapping_d2 = {
    'How much trust do you place in products that bear the KEBS label when it comes to quality and safety?': 'TrustKEBS',
    'Do you believe that products with the KEBS label are more reliable and safer than those without it?': 'BeliefKEBS'
}

# Rename the columns
d2 = d2.rename(columns=column_mapping_d2)

In [35]:
# Define your measurement model specifying the latent constructs and observed variables
model2 = '''
    # Measurement model
    f1 =~ TrustKEBS
    f2 =~ BeliefKEBS
'''

# Create a Model object with your specified model
model2 = semopy.Model(model2)

# Fit the CFA model to your data
fit = model2.fit(d2)

# Print the model fit results
print(fit)


Name of objective: MLW
Optimization method: SLSQP
Optimization successful.
Optimization terminated successfully
Objective value: 0.000
Number of iterations: 11
Params: 0.125 1.189 0.430 0.153 0.090


## CFA Results for Consumer Trust Construct

The objective of this analysis was to assess the Consumer Trust Construct using Confirmatory Factor Analysis (CFA). The analysis was conducted with the following parameters:

Name of Objective: MLW
Optimization Method: SLSQP
Optimization Status: Successful
Objective Value: 0.000
Number of Iterations: 11
Factor Loadings:

The factor loadings for each observed variable (survey questions) associated with the Consumer Trust Construct are as follows:

"How much trust do you place in products that bear the KEBS label when it comes to quality and safety?"

Factor Loading: 0.125
Interpretation: The factor loading of 0.125 suggests a weak positive association between this question and the Consumer Trust Construct. Individuals who place more trust in products with the KEBS label for quality and safety standards tend to have a slightly higher score on the Consumer Trust Construct. However, this question contributes to the measurement of the construct to a lesser extent.

"Do you believe that products with the KEBS label are more reliable and safer than those without it?"

Factor Loading: 1.189
Interpretation: The factor loading of 1.189 indicates a strong positive association between this question and the Consumer Trust Construct. Individuals who believe that products with the KEBS label are more reliable and safer tend to have a significantly higher score on the Consumer Trust Construct. This question is a robust indicator of consumer trust.

Conclusion:

In conclusion, the CFA results reveal that the question related to the belief in products with the KEBS label being more reliable and safer strongly contributes to the measurement of the Consumer Trust Construct. This question is a powerful predictor of consumer trust. On the other hand, the question about the extent of trust in KEBS-labeled products contributes to the construct but to a lesser degree, indicating a weaker association.

These findings are valuable for organizations and regulatory bodies aiming to understand and influence consumer trust in products labeled with KEBS standards




## Attitude Construct


In [36]:
column_mapping_d3 = {
    'Have you intentionally purchased products with the KEBS label because of its perceived quality and safety standards?': 'PurchaseKEBS',
    'To what extent do you associate the KEBS label with government oversight and regulation?': 'AssociationKEBS',
    'How likely are you to choose a product with the KEBS label over a similar product without it?': 'LikelihoodKEBS'
}

# Rename the columns
d3 = d3.rename(columns=column_mapping_d3)
d3


PurchaseKEBS  AssociationKEBS  LikelihoodKEBS
0               1                3               4
1               1                5               5
2               1                5               5
3               0                3               3
4               0                4               4
..            ...              ...             ...
110             1                5               5
111             1                5               5
112             1                4               5
113             1                4               5
114             1                5               5

[115 rows x 3 columns]

In [37]:

# Define your measurement model specifying the latent constructs and observed variables
model3 = '''
    # Measurement model
    f1 =~ PurchaseKEBS
    f2 =~ AssociationKEBS
    f3 =~ LikelihoodKEBS
'''

# Create a Model object with your specified model
model3 = semopy.Model(model3)

# Fit the CFA model to your data
fit = model3.fit(d3)
print(fit)

Name of objective: MLW
Optimization method: SLSQP
Optimization successful.
Optimization terminated successfully
Objective value: 0.000
Number of iterations: 15
Params: 0.933 0.996 0.081 0.090 0.212 0.232 0.345 0.708 0.365





## CFA Results for Consumer Attitude Construct

The objective of this analysis was to assess the Consumer Attitude Construct using Confirmatory Factor Analysis (CFA). The analysis was conducted with the following parameters:

Name of Objective: MLW
Optimization Method: SLSQP
Optimization Status: Successful
Objective Value: 0.000
Number of Iterations: 15
Factor Loadings:

The factor loadings for each observed variable (survey questions) associated with the Consumer Attitude Construct are as follows:

"Have you intentionally purchased products with the KEBS label because of its perceived quality and safety standards?"

Factor Loading: 0.933
Interpretation: The high factor loading of 0.933 suggests a strong positive association between this question and the Consumer Attitude Construct. Individuals who have intentionally purchased products with the KEBS label due to quality and safety perceptions are more likely to have a higher score on the Consumer Attitude Construct. This question is a robust indicator of consumer attitude.

"To what extent do you associate the KEBS label with government oversight and regulation?"

Factor Loading: 0.996
Interpretation: The factor loading of 0.996 indicates a very strong positive association between this question and the Consumer Attitude Construct. Individuals who strongly associate the KEBS label with government oversight and regulation tend to have a significantly higher score on the Consumer Attitude Construct. This question is a powerful predictor of consumer attitude.

"How likely are you to choose a product with the KEBS label over a similar product without it?"

Factor Loading: 0.081
Interpretation: The factor loading of 0.081, while positive, is notably lower compared to the other questions. This suggests that this question is only weakly associated with the Consumer Attitude Construct. It contributes to the measurement of the construct but to a much lesser extent compared to the first two questions.

Conclusion:

In conclusion, the CFA results indicate that the questions related to trust in products with the KEBS label and the association of KEBS with government oversight strongly contribute to the measurement of the Consumer Attitude Construct. These questions are robust indicators of consumer attitude and can be used effectively to assess the attitudes of consumers towards products labeled with KEBS standards. The third question, related to the likelihood of choosing KEBS-labeled products, contributes to the construct measurement but to a lesser degree.

These findings are essential for organizations and regulatory bodies to understand and influence consumer attitudes towards product quality and safety standards.